### PR TITLE
New data preprocessing

### PR DESCRIPTION
**Pull request by AvKhokhlov**

### CODE REVIEW

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 1]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
                                                    height,
                                                    input_channels);
         }
+
+        MLDenoiser::MLDenoiser(CLWContext context, Inference::Ptr inference, MLDenoiserInputs inputs)
+                   : m_inference(std::move(inference))
+        {
+            m_context = std::move(std::make_unique<CLWContext>(context));
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: `std::move` is not required. And in the next line.

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 2]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+            m_primitives = std::move(std::make_unique<CLWParallelPrimitives>(context));
+
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+            auto chanels_num = std::get<2>(shape);
+
+            size_t elems_count = sizeof(Tensor::ValueType) * width * height * chanels_num;
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: I think it's not _elem_count_, but may be _bytes_...

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 3]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+
+            m_device_cache = std::move(std::make_unique<CLWBuffer<char>>(
+                CLWBuffer<char>::Create(*m_context,
+                                        CL_MEM_READ_WRITE,
+                                        elems_count)));
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: on single line

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 4]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+        MLDenoiser::MLDenoiser(CLWContext context, Inference::Ptr inference, MLDenoiserInputs inputs)
+                   : m_inference(std::move(inference))
+        {
+            m_context = std::move(std::make_unique<CLWContext>(context));
+            m_primitives = std::move(std::make_unique<CLWParallelPrimitives>(context));
+
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+            auto chanels_num = std::get<2>(shape);
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: cha**nn**els_num

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 5]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+
+            // compute memory layout
+            switch (inputs)
+            {
+                case MLDenoiserInputs::kColorDepthNormalGloss7:
+                {
+                    m_layout.emplace(OutputType::kColor, 0u);
+                    m_layout.emplace(OutputType::kDepth, 3 * width * height);
+                    m_layout.emplace(OutputType::kViewShadingNormal, 4 * width * height);
+                    m_layout.emplace(OutputType::kGloss, 6 * width * height);
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: m_layout can be static initialized

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 6]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+            m_primitives = std::move(std::make_unique<CLWParallelPrimitives>(context));
+
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+            auto chanels_num = std::get<2>(shape);
+
+            size_t elems_count = sizeof(Tensor::ValueType) * width * height * chanels_num;
+
+            m_device_cache = std::move(std::make_unique<CLWBuffer<char>>(
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: Why do we need a unique_ptr here? Why not just CLWBuffer<char>?
std::move is not required in any case here.

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 7]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+                CLWBuffer<char>::Create(*m_context,
+                                        CL_MEM_READ_WRITE,
+                                        elems_count)));
+
+            m_host_cache = std::move(std::make_unique<std::uint8_t[]>(elems_count));
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: std::move not required here.

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 2 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: elem_count is CLW terminology for this type of buffers

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 3 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 1 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 4 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 8]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+                    }
+                    case Renderer::OutputType::kDepth:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kDepth];
+                        ProcessOutput<cl_float, Tensor::ValueType>(device_mem, mem_to_write);
+                        break;
+                    }
+                    case Renderer::OutputType::kViewShadingNormal:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kViewShadingNormal];
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: this line can ve place before switch - it;s common for all cases:
`auto mem_to_write = host_mem + m_layout[type];`

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 6 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: First, there is no necessary include in header file

Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Second, there is no suitable constructor for CLWBuffer

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 7 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 9]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+
+                        ProcessOutput<cl_float3, RadeonRays::float3>(device_mem,
+                            (Tensor::ValueType*)m_host_cache.get());
+
+                        // remove third chanel
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: it's better to write "copy only the first two channels".

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 10]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+                        break;
+                    }
+                    case Renderer::OutputType::kGloss:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kGloss];
+                        
+                        ProcessOutput<cl_float3, RadeonRays::float3>(device_mem,
+                            (Tensor::ValueType*)m_host_cache.get());
+
+                        // remove second and third chanels
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: it's better to write "copy only the first channel".

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 11]
```diff
@@ -22,5 +52,148 @@ namespace Baikal
...
+                        }
+                        break;
+                    }
+                }
+            }
+
+            m_inference->PushInput(std::move(tensor));
+            auto clw_output = dynamic_cast<ClwOutput*>(&output);
+
+            if (!clw_output)
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: I think this check should be earlier - in the begin of function

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 12]
```diff
@@ -22,5 +52,148 @@ namespace Baikal
...
+                        {
+                            mem_to_write[i / 3] = m_host_cache[i];
+                        }
+                        break;
+                    }
+                }
+            }
+
+            m_inference->PushInput(std::move(tensor));
+            auto clw_output = dynamic_cast<ClwOutput*>(&output);
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: lets rename

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 13]
```diff
@@ -22,5 +52,150 @@ namespace Baikal
...
+                        auto mem_to_write = host_mem + m_layout[OutputType::kDepth];
+                        ProcessOutput<cl_float, Tensor::ValueType>(device_mem, mem_to_write);
+                        break;
+                    }
+                    case Renderer::OutputType::kViewShadingNormal:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kViewShadingNormal];
+
+                        ProcessOutput<cl_float3, RadeonRays::float3>(device_mem,
+                            (Tensor::ValueType*)m_host_cache.get());
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: reinterpret_cast<>

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 14]
```diff
@@ -22,5 +52,148 @@ namespace Baikal
...
+
+            m_inference->PushInput(std::move(tensor));
+            auto clw_output = dynamic_cast<ClwOutput*>(&output);
+
+            if (!clw_output)
+            {
+                throw std::runtime_error("MLDenoiser::Apply(...): can not cast output");
+            }
+
+            // TODO: check empty
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: Done.

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 8 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 9 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 10 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.h [hunk 15]
```diff
@@ -1,11 +1,41 @@
...
 #include "PostEffects/ML/inference.h"
+#include "tensor.h"
 
 #include <cstddef>
 #include <memory>
 #include <string>
 
+class CLWContext;
+class CLWParallelPrimitives;
+class CLWOutput;
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.h]: Unused.

File: Baikal/PostEffects/ML/denoiser.h [hunk 16]
```diff
@@ -16,10 +46,30 @@ namespace Baikal
...
-                                                    float gpu_memory_fraction,
-                                                    std::string const& visible_devices,
-                                                    std::size_t width,
-                                                    std::size_t height);
+        class MLDenoiser : public PostEffect
+        {
+        public:
+
+            // Constructor
+            MLDenoiser(CLWContext context, Inference::Ptr inference, MLDenoiserInputs inputs);
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.h]: Forward declaration is not working here. To make it working, use CLWContext const&.

File: Baikal/PostEffects/ML/denoiser.h [hunk 17]
```diff
@@ -16,10 +46,30 @@ namespace Baikal
...
+
+            // Apply filter
+            void Apply(InputSet const& input_set, Output& output) override;
+
+        private:
+            using MemoryLayout = std::map<Renderer::OutputType, std::size_t>;
+
+            template <class ClType, class Type>
+            void ProcessOutput(CLWBuffer<RadeonRays::float3> input, Tensor::ValueType* host_mem);
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.h]: CLWBuffer const& (breaks forward declaration)

File: BaikalStandalone/Application/cl_render.cpp [hunk 18]
```diff
@@ -51,6 +51,12 @@ namespace Baikal
     {
         InitCl(settings, m_tex);
         LoadScene(settings);
+
+#ifdef ENABLE_ML_DENOISER
+        m_denoiser = std::make_unique<MLDenoiseProvider>(MLDenoiseProvider(
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: MLDenoiseProvider() call is excessive.

File: BaikalStandalone/Application/cl_render.h [hunk 19]
```diff
@@ -133,5 +138,10 @@ namespace Baikal
         //save GL tex for no interop case
         GLuint m_tex;
         Renderer::OutputType m_output_type;
+
+#ifdef ENABLE_ML_DENOISER
+        using MLDenoiseProviderPtr = std::unique_ptr<MLDenoiseProvider>;
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.h]: This does not save place.

File: Baikal/PostEffects/ML/denoiser.h [hunk 20]
```diff
@@ -21,5 +50,31 @@ namespace Baikal
                                                     std::string const& visible_devices,
                                                     std::size_t width,
                                                     std::size_t height);
+
+        class MLDenoiser : public PostEffect
+        {
+        public:
+
+            // Constructor
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.h]: Useless.

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 21]
```diff
@@ -22,5 +52,146 @@ namespace Baikal
...
+            auto height = std::get<1>(shape);
+
+            auto tensor = m_inference->GetInputTensor();
+            auto host_mem = tensor.data();
+
+            for (const auto& input : input_set)
+            {
+                auto type = input.first;
+                auto clw_output = static_cast<ClwOutput*>(input.second);
+                auto mem_to_write = host_mem;
```
Comment (achernigin1987) [inline Baikal/PostEffects/ML/denoiser.cpp]: I think, it will not work...

File: BaikalStandalone/Utils/config_manager.h [hunk 22]
```diff
@@ -33,53 +33,45 @@ namespace Baikal
...
-        std::vector<Config>& renderers,
-        int initial_num_bounces,
-        int req_platform_index = -1,
-        int req_device_index = -1);
+    Config() = default;
 
-private:
+    Config(Config&& cfg) = default;
 
+    ~Config()
```
Comment (dtarakanov1) [inline BaikalStandalone/Utils/config_manager.h]: What for?

File: BaikalStandalone/Utils/config_manager.h [hunk 23]
```diff
@@ -33,53 +33,45 @@ namespace Baikal
...
+    bool caninterop;
 
-    static void CreateConfigs(
-        Mode mode,
-        bool interop,
-        std::vector<Config>& renderers,
-        int initial_num_bounces,
-        int req_platform_index = -1,
-        int req_device_index = -1);
+    Config() = default;
```
Comment (dtarakanov1) [inline BaikalStandalone/Utils/config_manager.h]: Remove?

File: BaikalStandalone/Utils/config_manager.h [hunk 24]
```diff
@@ -33,53 +33,45 @@ namespace Baikal
...
 
-private:
+    Config(Config&& cfg) = default;
```
Comment (dtarakanov1) [inline BaikalStandalone/Utils/config_manager.h]: Remove?

File: BaikalStandalone/Utils/config_manager.h [hunk 25]
```diff
@@ -33,53 +33,45 @@ namespace Baikal
...
 
+void CreateConfigs(
+    Mode mode,
+    bool interop,
+    std::vector<Config>& renderers,
+    int initial_num_bounces,
+    int req_platform_index = -1,
+    int req_device_index = -1);
+
 #endif // CONFIG_MANAGER_H
```
Comment (dtarakanov1) [inline BaikalStandalone/Utils/config_manager.h]: Having that almost whole file is rewritten, let's switch to pragma once.

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 14 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.h (hunk 17 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.h]: Done

File: BaikalStandalone/Application/cl_render.cpp (hunk 18 above)
Comment (AvKhokhlov) [reply BaikalStandalone/Application/cl_render.cpp]: Done

File: BaikalStandalone/Application/cl_render.h (hunk 19 above)
Comment (AvKhokhlov) [reply BaikalStandalone/Application/cl_render.h]: Done

File: BaikalStandalone/Utils/config_manager.h (hunk 22 above)
Comment (AvKhokhlov) [reply BaikalStandalone/Utils/config_manager.h]: Done

File: BaikalStandalone/Utils/config_manager.h (hunk 23 above)
Comment (AvKhokhlov) [reply BaikalStandalone/Utils/config_manager.h]: Done

File: BaikalStandalone/Utils/config_manager.h (hunk 24 above)
Comment (AvKhokhlov) [reply BaikalStandalone/Utils/config_manager.h]: Nope

File: BaikalStandalone/Utils/config_manager.h (hunk 25 above)
Comment (AvKhokhlov) [reply BaikalStandalone/Utils/config_manager.h]: Done

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 26]
```diff
@@ -22,5 +52,142 @@ namespace Baikal
...
+        void MLDenoiser::Apply(InputSet const& input_set, Output& output)
+        {
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+
+            auto tensor = m_inference->GetInputTensor();
+            auto host_mem = tensor.data();
+
+            for (const auto& input : input_set)
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: This code depends on the inputs iteration order which in the case of std::map is the `Renderer::OutputType`  values order. Instead we should loop through m_layout (m_layout should be something like `std::vector<std::pair<Renderer::OutputType, size_t>>`) the and get corresponding inputs with `input_set.at(type)`.

**Comment by dtarakanov1** [comment]
The code was copied into the `integration` branch, which then was merged to the `master` branch.

**Pull request closed without merging**